### DISCUSSION TITLE
How not to directly depend on Pydantic in inner layers?

### DISCUSSION THREAD

**Discussion by ydennisy** [pydantic/pydantic#4729] [Question]
👋 hey!

I love using pydantic with fastapi. I am trying to build clean architecture using domain driven design.

In reality my main goal is not to allow 3rd partly libs to be direct dependencies of the inner layers of my system.

Since we are defining types using pydantic right at the outer layer it than feels like a good idea to just import this type across all the other layers when needed.

Is there a way to avoid this? Using sometime like dependency inversion or create some more generic interface?

Thanks!!!

EDIT - asked for an example in the comments.

```python

# routes.py

class Book(BaseModel):
  isbn: str

@app.post('/books')
def post_books(payload: Book):
  use_case.execute(payload)
```

So this file is all good! We have our later layer of the system and it depends on pydantic BaseModel.

Now we write the code for our use case (inner business logic).

```python

# add_book.py

from routes import Book # this is the issue! Whilst very nice and easy, it means we now depend on pydantic in the inner layers of the system.

class AddBook:

  def execute(payload: Book):
    # do business logic
```

Hopefully that helps? 

The issue is that the inner business logic layers should not depend directly on `BaseModel` as this could be replaced in future.

**Comment by samuelcolvin** [comment]
I'm not that clear in the question. Might help to add some code to illustrate what you're asking?

**Reply by ydennisy** [reply]
@samuelcolvin I have had a go at adding some code to help illustrate the question, please let me know if that helps :)

**Comment by tuchandra** [comment]
If I understand correctly, you're looking for a way to express that your `execute` business-layer function needs a `Book`-like object without actually using the `Book` class. If that's not the case, I apologize.

One option is to guard the import in `add_book.py` with an `if typing.TYPE_CHECKING` clause. This would mean that it's not imported at runtime, but you can still check the code statically with mypy/pyright or in your IDE. If that's still too much dependence, and you're not looking to import the object at all, then you'd probably want to define a [Protocol](https://peps.python.org/pep-0544/) to specify the interface that your function expects—

```
from typing import Protocol

class BookProto(Protocol):
  isbn: str

def execute(payload: BookProto):
  ...
```

This can be cumbersome if you have a lot of models or functions, but it's probably the safest way to express that you want "something that has an `isbn: str` attribute" without requiring the rest of the model. This would allow you to pass the Book model in your route, but (probably, I haven't tried this) also allow a dataclass with `isbn: str` if you strip out pydantic entirely.

**Comment by alexlopezc** [comment]
Did you find a solution? I have a similar problem and I do not like the domain to depend on the model from the routes (BaseModel)

**Reply by simon376** [reply]
did you? same issue here

**Reply by alexlopezc** [reply]
I didn't but the models from the routes may be different from the ones from the domain. Even you may have different similar models in the routes (User with password or without, or partially filled). There are some examples in FastAPI docs about this.

In addition, a change in the API will force a change in the domain if you reuse the same model. So I would avoid as much as possible to have that dependency.

You can have something like a builder/factory between route and domain models but I would like to find examples of someone else doing it...

I'm not currently building anything similar so I will have to take a look at what others are recommending.

**Comment by simon376** [comment]
How to convert between inner and outer models without too much boilerplate
or code that needs to be maintained is still unclear for me, like the
builder/factory you mentioned.. Thanks anyway!

Alex Lopez ***@***.***> schrieb am Do., 12. Dez. 2024, 16:24:

**Comment by ErikvdVen** [comment]
I stumbled upon the same challenge: you want to use POPO's (Plain Old Python Objects) in your domain. I have seen quite a couple of DDD examples in Python, but almost all of them use Pydantic in the domain and that's a problem IMHO. Pydantic is great, don't get me wrong, amazing for data serialization but you don't want to use it in your domain layer.

I did quite some research as I don't want to create Object Mappers myself, but it is something you want to use: You want a mapper that is capable of mapping Entities to dictionaries in your Application layer or the other way around.

I just stumbled upon this library: https://github.com/konradhalas/dacite And it seems to me this is exactly what I need! I'll update this answer after I've tried the package for a while. Maybe not the best place to talk about different libraries, but I think it helps people who consider using Pydantic but don't want to end up with Pydantic in the domain layer, neither.

**Reply by ErikvdVen** [reply]
I promised a quick update, so here it is. I'm very satisfied with Dacite so far. It helps with the mapping between Python dataclasses and Pydantic Basemodels by reducing some boilerplate when writing mappers, which saves me work and makes the code look cleaner, something I'm quite happy about. People who prefer to avoid third-party libraries will probably opt for explicit mapping instead.

I’ve written a blog article explaining how I use Dacite in some of the mappers between my repository layer and domain layer for a personal side project: https://coderik.nl/posts/keep-pydantic-out-of-your-domain-layer/ 

I don’t cover endpoints in the article, but you could apply mappers there as well. And if you'd rather not call a mapper directly in your router function body, you can also use an Annotated together with Depends:

```python
from fastapi import FastAPI, Depends
from pydantic import BaseModel
from dataclasses import dataclass
from typing import Annotated, TypeVar, Callable

from dacite import from_dict

app = FastAPI()

class CreateProjectModel(BaseModel):
    project_name: str
    project_description: str

class Entity:
    ...

@dataclass
class Project(Entity):
    project_name: str
    project_description: str


MapperEntity = TypeVar("MapperEntity", bound=Entity)
MapperModel = TypeVar("MapperModel", bound=BaseModel)

def mapper_factory(from_model: type[MapperModel], to_entity: type[MapperEntity]) -> Callable[[MapperModel], MapperEntity]:
    def _dependency(body: from_model) -> MapperEntity:
        return from_dict(data_class=to_entity, data=body.model_dump())
    return _dependency

@app.post("/projects")
def create_project(project: Annotated[Project, Depends(mapper_factory(CreateProjectModel, Project))]):
    return {
        "message": "Project created successfully",
        "project": {
            "name": project.project_name,
            "description": project.project_description,
        }
    }
```

I should add that I didn’t create the DataMapper classes I describe in the [article](https://coderik.nl/posts/keep-pydantic-out-of-your-domain-layer/ ) just to enforce a clearer separation of concerns. I also did it for practical reasons, as it's easier to perform some explicit mappings this way, which may vary per use case. For example, converting strings into specific objects before passing the rest of the keys to Dacite:

```python
class RecipeDataMapper(AbstractDataMapper[Recipe, dict]): 

    ... # simplified for brevity

    def model_to_entity(self, instance: dict) -> Recipe:
        instance["start_time"] = datetime.fromisoformat(instance["start_time"])
        return from_dict(data_class=self.entity_class, data=instance)

    ... # simplified for brevity

```
